src/lib/transition_handler/breadcrumb_builder.ml

```diff
@@ -0,0 +1,98 @@
+        let%map subtree_of_constructions =
+          Rose_tree.Deferred.Or_error.fold_map subtree_of_transitions
+            ~init:(Cached.pure `Initial)
+            ~f:(fun cached_parent_or_initial cached_transition ->
+              let open Deferred.Let_syntax in
+              let%map cached_result =
+                Cached.transform cached_transition ~f:(fun transition ->
+                    let open Deferred.Or_error.Let_syntax in
+                    let parent_or_initial =
+                      Cached.peek cached_parent_or_initial
+                    in
+                    let%bind well_formed_parent =
+                      match parent_or_initial with
+                      | `Initial -> breadcrumb_if_present () |> Deferred.return
+                      | `Constructed parent -> Deferred.Or_error.return parent
+                    in
+                    let expected_parent_hash =
+                      Transition_frontier.Breadcrumb.transition_with_hash
+                        well_formed_parent
+                      |> With_hash.hash
+                    in
+                    let actual_parent_hash =
+                      transition |> With_hash.data
+                      |> External_transition.Verified.protocol_state
+                      |> External_transition.Protocol_state.previous_state_hash
+                    in
+                    let%bind () =
+                      Deferred.return
+                        (Result.ok_if_true
+                           (State_hash.equal actual_parent_hash
+                              expected_parent_hash)
+                           ~error:
+                             (Error.of_string
+                                "Previous external transition hash does not \
+                                 equal to current external transition's \
+                                 parent hash"))
+                    in
+                    let open Deferred.Let_syntax in
+                    match%map
+                      Transition_frontier.Breadcrumb.build ~logger
+                        ~parent:well_formed_parent
+                        ~transition_with_hash:transition
+                    with
+                    | Ok new_breadcrumb ->
+                        let open Result.Let_syntax in
+                        let%map _ : Transition_frontier.Breadcrumb.t =
+                          breadcrumb_if_present ()
+                        in
+                        `Constructed new_breadcrumb
+                    | Error (`Fatal_error exn) -> Or_error.of_exn exn
+                    | Error (`Validation_error error) -> Error error )
+                |> Cached.sequence_deferred
+              in
+              Cached.sequence_result cached_result )
+        in
+        Rose_tree.map subtree_of_constructions ~f:(fun construction ->
+            Cached.transform construction ~f:(function
+              | `Initial -> failwith "impossible"
```

Comment:
Is it possible to write this function without `Initial. It doesn't feel necessary and you are left with this failwith in the end.